### ISSUE TITLE
typings for the route config based hooks no longer working since 4.16.0

### ISSUE THREAD

**Issue by vidarc** [fastify/fastify#4705]
### Prerequisites

- [X] I have written a descriptive issue title
- [X] I have searched existing issues to ensure the bug has not already been reported

### Fastify version

4.16.0

### Node.js version

18.x

### Operating system

macOS

### Operating system version (i.e. 20.04, 11.3, 10)

12.3

### Description

The hooks specified in the options of a [route](https://www.fastify.io/docs/latest/Reference/Routes/) no longer have the correct typings applied/resolved. They come back as `any`. The following code snippet works in `4.15.0`, but breaks in `4.16.0`

### Steps to Reproduce

```typescript
import Fastify from "fastify";

const fastify = Fastify();

fastify.get(
  "/route",
  {
    onError: (request, reply, error) => {
      request // is any
      reply // is any
      error // is any
    },
  },
  async () => ({ success: "yes" })
);
```
The same appears to be true for all the other hooks specified in this manner as well.

Interestingly, hooks added with `addHook` appear to have the correct typings resolved

**Comment by Eomm** [comment]
Thanks for reporting!
Would you like to send a Pull Request to address this issue? Remember to add unit tests.

**Comment by mcollina** [comment]
Can you please include a script that compiles in v4.15.0 and it does not in v4.16.0. That would help in bisecting the problem.

**Comment by vidarc** [comment]
Do you need something more than the attached code snippet? That shows the problem between the 2 versions

**Comment by vidarc** [comment]
It's definitely due to https://github.com/fastify/fastify/pull/4655

If I take the `types/route.d.ts` file from before that PR was merged, the types work as expected

**Comment by mcollina** [comment]
Anything that I can quickly run in a bash shell that returns 1 if it fails. Opening an editor and trying to autocomplete is not great for git bisecting.

**Comment by Fdawgs** [comment]
closed by #4708
